### PR TITLE
Implement link checker in Github Actions (CI)

### PR DESCRIPTION
**Pull request by rejahrehim**
This PR covers issue  #300 #301 #302  #304 #305

- [x] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

**What did this PR accomplish?**

- Implement lint checker to validate the links in each PR

### CODE REVIEW

File: .github/workflows/md-link-check.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    - name: Setup Node
+      uses: actions/setup-node@v1
+      with:
+        node-version: 12.x
+    - name: Install dependencies
+      run: npm install -g markdown-link-check
+    - name: Run link check
+      run: npm run-script link-check
+    - name: Show broken links
+      id: broknlinks
```
Comment (kingthorin) [inline .github/workflows/md-link-check.yml]: Is this mispelt on purpise?

File: .github/workflows/md-link-check.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        links=`cat brokenlinks`
+        links="${links//'%'/'%25'}"
+        links="${links//$'\n'/'%0A'}"
+        links="${links//$'\r'/'%0D'}"
+        echo ::set-output name=links::**Following links are broken:** %0A$links
+    - name: Send comment to PR with broken links
+      if: failure() && github.event_name == 'pull_request'
+      uses: thollander/actions-comment-pull-request@master
+      with:
+        message: ${{ steps.broknlinks.outputs.links }}
```
Comment (kingthorin) [inline .github/workflows/md-link-check.yml]: Same, re: spelling

File: .github/workflows/md-link-check.yml (hunk 1 above)
Comment (rejahrehim) [reply .github/workflows/md-link-check.yml]: Typo :(

File: .github/workflows/md-link-check.yml (hunk 2 above)
Comment (rejahrehim) [reply .github/workflows/md-link-check.yml]: Typo :(

File: package.json [hunk 3]
```diff
@@ -8,7 +8,8 @@
     "markdownlint-cli": "^0.19.0"
   },
   "scripts": {
-    "test": "markdownlint .  --ignore node_modules"
+    "test": "markdownlint .  --ignore node_modules",
+    "link-check" : "find  document -name \\*.md -exec markdown-link-check -c markdown-link-check-config.json 1> log 2> err {} \\; && if [ -e err ] && grep -q  \"ERROR:\" err ; then exit 113  ; else echo -e \"All good\"; fi"
```
Comment (victoriadrake) [inline package.json]: "All good" is non-specific feedback; I'd prefer "No broken links found."

Comment (victoriadrake) [reply package.json]: What does `113` typically indicate, besides being non-zero?

Comment (rejahrehim) [reply package.json]: `113` Just a user-defined exit code, any other recommendation?

Comment (rejahrehim) [reply package.json]: > "All good" is non-specific feedback; I'd prefer "No broken links found."

Updated

**Comment by kingthorin** [comment]
I guess this will continue to fail until rebased with all fixed links?

**Comment by rejahrehim** [comment]
> I guess this will continue to fail until rebased with all fixed links?

Yes. We have to resolve those broken links anyway.

**Comment by rejahrehim** [comment]
**Current broken links:** 

**FILE:** 4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.7_Map_Execution_Paths_Through_Application_OTG-INFO-007.md
[✖] https://en.wikipedia.org/wiki/Diagramming_software → Status: 404


**FILE:** 4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.14.3_Testing_for_Format_String.md
[✖] http://comsec.theclerk.com/CISSP/FormatString.pdf → Status: 404
[✖] http://www.cs.ucsb.edu/~jzhou/security/formats-teso.html → Status: 404
[✖] http://julianor.tripod.com/format-bug-analysis.pdf → Status: 404

**FILE:** 4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.6_Testing_for_LDAP_Injection_OTG-INPVAL-006.md
[✖] https://en.wikipedia.org/wiki/Polish_notation_prefix_notation → Status: 404

**FILE:** 4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.15_Testing_for_Incubated_Vulnerability_OTG-INPVAL-015.md
[✖] http://lists.grok.org.uk/pipermail/full-disclosure/2006-July/048059.html → Status: 404

**FILE:** 2_Introduction/2_Introduction.md
[✖] https://www.archive.ece.cmu.edu/~ganger/712.fall02/papers/p761-thompson.pdf → Status: 0

**Comment by rejahrehim** [comment]
**Current broken links:**


FILE:document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.7_Map_Execution_Paths_Through_Application_WSTG-INFO-007.md
[✖] https://en.wikipedia.org/wiki/Diagramming_software → Status: 404

FILE:document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.2_Testing_for_MySQL.md
[✖] https://www.owasp.org/index.php/Testing_for_SQL_Injection_(WSTG-INPVAL-005) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.4_OWASP_Backend_Security_Project_Testing_PostgreSQL.md
[✖] https://www.owasp.org/index.php/Testing_for_SQL_Injection_(WSTG-INPVAL-005) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.14.3_Testing_for_Format_String.md
[✖] http://comsec.theclerk.com/CISSP/FormatString.pdf → Status: 404
[✖] http://www.cs.ucsb.edu/~jzhou/security/formats-teso.html → Status: 404
[✖] http://julianor.tripod.com/format-bug-analysis.pdf → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.15_Testing_for_Incubated_Vulnerability_WSTG-INPVAL-015.md
[✖] https://www.owasp.org/index.php/Testing_for_Stored_Cross_site_scripting_(WSTG-INPVAL-002) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_SQL_Injection_(WSTG-INPVAL-006) → Status: 404
[✖] http://lists.grok.org.uk/pipermail/full-disclosure/2006-July/048059.html → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.6_Testing_for_NoSQL_Injection.md
[✖] https://www.owasp.org/index.php/Testing_for_HTTP_Parameter_pollution_(WSTG-INPVAL-004) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.11_Business_Logic_Testing/4.11.2_Test_Ability_to_Forge_Requests_WSTG-BUSLOGIC-002.md
[✖] https://www.owasp.org/index.php/Testing_for_Exposed_Session_Variables_(WSTG-SESS-004) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_CSRF_(WSTG-SESS-005) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_Account_Enumeration_and_Guessable_User_Account_(WSTG-IDENT-004) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.11_Business_Logic_Testing/4.11.8_Test_Upload_of_Unexpected_File_Types_WSTG-BUSLOGIC-008.md
[✖] https://www.owasp.org/index.php/Test_File_Extensions_Handling_for_Sensitive_Information_(WSTG-CONFIG-003) → Status: 404
[✖] https://www.owasp.org/index.php/Test_Upload_of_Malicious_Files_(WSTG-BUSLOGIC-009) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.11_Business_Logic_Testing/4.11.6_Testing_for_the_Circumvention_of_Work_Flows_WSTG-BUSLOGIC-006.md
[✖] https://www.owasp.org/index.php/Testing_Directory_traversal/file_include_(WSTG-AUTHZ-001) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_Bypassing_Authorization_Schema_(WSTG-AUTHZ-002) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_Session_Management_Schema_(WSTG-SESS-001) → Status: 404
[✖] https://www.owasp.org/index.php/Test_business_logic_data_validation_(WSTG-BUSLOGIC-001) → Status: 404
[✖] https://www.owasp.org/index.php/Test_Ability_to_forge_requests_(WSTG-BUSLOGIC-002) → Status: 404
[✖] https://www.owasp.org/index.php/Test_integrity_checks_(WSTG-BUSLOGIC-003) → Status: 404
[✖] https://www.owasp.org/index.php/Test_for_Process_Timing_(WSTG-BUSLOGIC-004) → Status: 404
[✖] https://www.owasp.org/index.php/Test_number_of_times_a_function_can_be_used_limits_(WSTG-BUSLOGIC-005) → Status: 404
[✖] https://www.owasp.org/index.php/Test_defenses_against_application_mis-use_(WSTG-BUSLOGIC-007) → Status: 404

[✖] https://www.owasp.org/index.php/Test_Upload_of_Unexpected_File_Types_(WSTG-BUSLOGIC-008) → Status: 404
[✖] https://www.owasp.org/index.php/Test_Upload_of_Malicious_Files_(WSTG-BUSLOGIC-009) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.11_Business_Logic_Testing/4.11.1_Test_Business_Logic_Data_Validation_WSTG-BUSLOGIC-001.md
[✖] https://www.owasp.org/index.php/Testing_for_Account_Enumeration_and_Guessable_User_Account_(WSTG-IDENT-004) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_Session_Management_Schema_(WSTG-SESS-001) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_Exposed_Session_Variables_(WSTG-SESS-004) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.10_Testing_for_Weak_Cryptography/4.10.3_Testing_for_Sensitive_Information_Sent_via_Unencrypted_Channels_WSTG-CRYPST-003.md
[✖] https://www.owasp.org/index.php/Testing_for_cookies_attributes_(WSTG-SESS-002) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.10_Testing_for_Weak_Cryptography/4.10.1_Testing_for_Weak_SSL_TLS_Ciphers_Insufficient_Transport_Layer_Protection_WSTG-CRYPST-001.md
[✖] https://www.owasp.org/index.php/Testing_for_Sensitive_information_sent_via_unencrypted_channels_(WSTG-CRYPST-003) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_Credentials_Transported_over_an_Encrypted_Channel_(WSTG-AUTHN-001) → Status: 404
[✖] https://www.owasp.org/index.php/Test_Network/Infrastructure_Configuration_(WSTG-CONFIG-001) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_cookies_attributes_(WSTG-SESS-002) → Status: 404
[✖] https://www.owasp.org/index.php/Test_HTTP_Strict_Transport_Security_(WSTG-CONFIG-007) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.9_Testing_for_Error_Handling/4.9.1_Testing_for_Error_Code_WSTG-ERR-001.md
[✖] https://www.owasp.org/index.php/Conduct_search_engine_discovery/reconnaissance_for_information_leakage_(WSTG-INFO-001) → Status: 404
[✖] https://www.owasp.org/index.php/Fingerprint_Web_Server_(WSTG-INFO-002) → Status: 404
[✖] https://www.owasp.org/index.php/Testing_for_SQL_Injection_(WSTG-INPVAL-005) → Status: 404

FILE:document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.7_Testing_for_ORM_Injection_WSTG-INPVAL-007.md
[✖] https://www.owasp.org/index.php/Testing_for_SQL_Injection_(WSTG-INPVAL-005)#References → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.11_Business_Logic_Testing/4.11.9_Test_Upload_of_Malicious_Files_WSTG-BUSLOGIC-009.md
[✖] https://www.owasp.org/index.php/Test_File_Extensions_Handling_for_Sensitive_Information_(WSTG-CONFIG-003) → Status: 404
[✖] https://www.owasp.org/index.php/Test_Upload_of_Unexpected_File_Types_(WSTG-BUSLOGIC-008) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.11_Business_Logic_Testing/4.11.4_Test_for_Process_Timing_WSTG-BUSLOGIC-004.md
[✖] https://www.owasp.org/index.php/Testing_for_cookies_attributes_(WSTG-SESS-002) → Status: 404
[✖] https://www.owasp.org/index.php/Test_Session_Timeout_(WSTG-SESS-007) → Status: 404


FILE:document/4_Web_Application_Security_Testing/4.12_Client_Side_Testing/4.12.10_Testing_WebSockets_WSTG-CLIENT-010.md
[✖] https://www.owasp.org/index.php/Testing_for_Weak_SSL/TLS_Ciphers,_Insufficient_Transport_Layer_Protection_(WSTG-CRYPST-001) → Status: 404

**Comment by kingthorin** [comment]
Wikipedia?

**Comment by kingthorin** [comment]
Something isn't quite right with this, there are definitely things there that load fine manually.

**Comment by rejahrehim** [comment]
> Something isn't quite right with this, there are definitely things there that load fine manually.

I have checked that in the previous run. As of now, many are OWASP links.

**Comment by rejahrehim** [comment]
> Wikipedia?

https://en.wikipedia.org/wiki/Diagramming_software

This link in 404. 

Screenshot 2020-02-14 at 1 22 11 PM

**Comment by kingthorin** [comment]
> This link in 404.

Ha, you're right, I glanced at it on a cell, then also one for a full disclosure list. They both seem like content on a small screen at a glance but are actual 404s. Thanks & sorry. My bad :cry:

**Comment by kingthorin** [comment]
The OWASP ones are all broken because of the OTG > WSTG change, but shoukd probably link locally anyway. I'll go through them.

**Comment by kingthorin** [comment]
Suggestions (if we're in agreement on these I'll open a PR):

> FILE: 4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.7_Map_Execution_Paths_Through_Application_OTG-INFO-007.md
> [✖] https://en.wikipedia.org/wiki/Diagramming_software → Status: 404

✔️ https://en.wikipedia.org/wiki/List_of_concept-_and_mind-mapping_software

> FILE: 4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.14.3_Testing_for_Format_String.md
> [✖] http://comsec.theclerk.com/CISSP/FormatString.pdf → Status: 404

✔️ https://seclists.org/bugtraq/2000/Sep/214

> [✖] http://www.cs.ucsb.edu/~jzhou/security/formats-teso.html → Status: 404

✔️ https://cs155.stanford.edu/papers/formatstring-1.2.pdf

> [✖] http://julianor.tripod.com/format-bug-analysis.pdf → Status: 404

✔️ http://julianor.tripod.com/bc/format-bug-analysis.pdf

> FILE: 4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.15_Testing_for_Incubated_Vulnerability_OTG-INPVAL-015.md
> [✖] http://lists.grok.org.uk/pipermail/full-disclosure/2006-July/048059.html → Status: 404

✔️ https://cxsecurity.com/issue/WLB-2006080004

**Comment by rejahrehim** [comment]
> Suggestions (if we're in agreement on these I'll open a PR):
> 
> > FILE: 4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.7_Map_Execution_Paths_Through_Application_OTG-INFO-007.md
> > [✖] https://en.wikipedia.org/wiki/Diagramming_software → Status: 404
> 
> ✔️ https://en.wikipedia.org/wiki/List_of_concept-_and_mind-mapping_software
> 
> > FILE: 4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.14.3_Testing_for_Format_String.md
> > [✖] http://comsec.theclerk.com/CISSP/FormatString.pdf → Status: 404
> 
> ✔️ https://seclists.org/bugtraq/2000/Sep/214
> 
> > [✖] http://www.cs.ucsb.edu/~jzhou/security/formats-teso.html → Status: 404
> 
> ✔️ https://cs155.stanford.edu/papers/formatstring-1.2.pdf
> 
> > [✖] http://julianor.tripod.com/format-bug-analysis.pdf → Status: 404
> 
> ✔️ http://julianor.tripod.com/bc/format-bug-analysis.pdf
> 
> > FILE: 4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.15_Testing_for_Incubated_Vulnerability_OTG-INPVAL-015.md
> > [✖] http://lists.grok.org.uk/pipermail/full-disclosure/2006-July/048059.html → Status: 404
> 
> ✔️ https://cxsecurity.com/issue/WLB-2006080004

Looks good for me, still one http link `http://julianor.tripod.com/bc/format-bug-analysis.pdf`

**Comment by kingthorin** [comment]
Sadly there does not seem to be a TLS enabled alternative for that one.

**Comment by kingthorin** [comment]
Once https://github.com/OWASP/wstg/pull/327 is approved and merge I'll tackle a new PR for these five.

**Pull request merged**